uvicorn/protocols/websockets/wsproto_impl.py

```diff
@@ -125,9 +125,12 @@ def resume_writing(self):
         self.writable.set()
 
     def shutdown(self):
-        self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
-        output = self.conn.send(wsproto.events.CloseConnection(code=1012))
-        self.transport.write(output)
+        if self.handshake_complete:
+            self.queue.put_nowait({"type": "websocket.disconnect", "code": 1012})
+            output = self.conn.send(wsproto.events.CloseConnection(code=1012))
+            self.transport.write(output)
+        else:
+            self.send_500_response()
```

Comment:
The thing is... If the above is `False`, we still need to send something to the client, otherwise we are being rude. :disappointed: 

Without these lines, the client will have a:
```bash
curl: (52) Empty reply from server
```

With these lines:
```
HTTP/1.1 500 
content-length: 0
```

The curl command used was:
```bash
curl --include \
     --no-buffer \
     --header "Connection: Upgrade" \
     --header "Upgrade: websocket" \
     --header "Host: example.com:80" \
     --header "Origin: http://example.com:80" \
     --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" \
     --header "Sec-WebSocket-Version: 13" \
     http://localhost:8000/
```

Reply:
Much neater. 😌

I did spend some time trying to figure out if we should we also include a textual description here, but I think it's probably okay as it currently stands.